File: entities/player.py
"""玩家坦克。"""
import pygame

import config
from entities.tank import Tank
from utils.helpers import load_image


class PlayerTank(Tank):
    """由玩家控制的坦克。"""

    def __init__(self, x, y):
        super().__init__(x, y, "player_tank.png", config.PLAYER_SPEED, config.PLAYER_SHOOT_COOLDOWN)
        self.lives = config.PLAYER_LIVES
        self.invincible_timer = 0.0
        self.respawn_timer = 0.0
        self.spawn_pos = (x, y)
        self._blink_timer = 0.0
        self._visible = True

    def handle_input(self, keys):
        """处理键盘输入。"""
        if self.respawn_timer > 0:
            return
        moved = False
        if keys[pygame.K_UP] or keys[pygame.K_w]:
            self.set_direction(config.DIR_UP)
            moved = True
        elif keys[pygame.K_DOWN] or keys[pygame.K_s]:
            self.set_direction(config.DIR_DOWN)
            moved = True
        elif keys[pygame.K_LEFT] or keys[pygame.K_a]:
            self.set_direction(config.DIR_LEFT)
            moved = True
        elif keys[pygame.K_RIGHT] or keys[pygame.K_d]:
            self.set_direction(config.DIR_RIGHT)
            moved = True
        self.moving = moved

    def update(self, dt, level, enemies):
        """更新玩家状态。"""
        super().update(dt)
        if self.respawn_timer > 0:
            self.respawn_timer -= dt
            if self.respawn_timer <= 0:
                self.respawn(self.spawn_pos[0], self.spawn_pos[1])
            return
        if self.invincible_timer > 0:
            self.invincible_timer -= dt
            self._blink_timer += dt
            if self._blink_timer >= 0.1:
                self._blink_timer = 0.0
                self._visible = not self._visible
        else:
            self._visible = True
        if self.moving:
            # 玩家可以穿过敌人坦克，但接触时会受伤（在 PlayState 中处理）
            self.try_move(dt, level, [])

    def draw(self, surface):
        if not self._visible:
            return
        super().draw(surface)

    def shoot(self):
        """玩家射击。"""
        if self.respawn_timer > 0:
            return None
        return super().shoot("player")

    def take_damage(self):
        """玩家受击：扣生命并进入复活。"""
        if self.respawn_timer > 0 or self.invincible_timer > 0:
            return
        self.lives -= 1
        if self.lives > 0:
            self.respawn_timer = config.PLAYER_RESPAWN_TIME
            self.invincible_timer = 0.0
            self._visible = False
        else:
            self.kill()

    def respawn(self, x, y):
        """复活。"""
        super().respawn(x, y)
        self.invincible_timer = config.PLAYER_INVINCIBLE_TIME
        self._visible = True
        self._blink_timer = 0.0

    def is_dead(self):
        """生命耗尽且未在复活中。"""
        return self.lives <= 0 and self.respawn_timer <= 0

File: js/entities/player.js
"use strict";

var CONFIG = require('../config');
var Tank = require('./tank');

function PlayerTank(x, y, assets) {
  var asset = function (ctx, dx, dy, size) {
    assets.drawPlayerTank(ctx, dx, dy, size);
  };
  Tank.call(this, x, y, asset, CONFIG.PLAYER_SPEED, CONFIG.PLAYER_SHOOT_COOLDOWN);
  this.lives = CONFIG.PLAYER_LIVES;
  this.invincibleTimer = 0;
  this.respawnTimer = 0;
  this.spawnPos = { x: x, y: y };
  this.blinkTimer = 0;
  this.visible = true;
}

PlayerTank.prototype = Object.create(Tank.prototype);
PlayerTank.prototype.constructor = PlayerTank;

PlayerTank.prototype.handleInput = function (input) {
  if (this.respawnTimer > 0) {
    return;
  }
  if (input.dpadDirection !== null) {
    this.setDirection(input.dpadDirection);
    this.moving = true;
  } else {
    this.moving = false;
  }
};

PlayerTank.prototype.update = function (dt, level, enemies) {
  Tank.prototype.update.call(this, dt);
  if (this.respawnTimer > 0) {
    this.respawnTimer -= dt;
    if (this.respawnTimer <= 0) {
      this.respawn(this.spawnPos.x, this.spawnPos.y);
    }
    return;
  }
  if (this.invincibleTimer > 0) {
    this.invincibleTimer -= dt;
    this.blinkTimer += dt;
    if (this.blinkTimer >= 0.1) {
      this.blinkTimer = 0;
      this.visible = !this.visible;
    }
  } else {
    this.visible = true;
  }
  if (this.moving) {
    this.tryMove(dt, level, []);
  }
};

PlayerTank.prototype.draw = function (ctx) {
  if (!this.visible) {
    return;
  }
  Tank.prototype.draw.call(this, ctx);
};

PlayerTank.prototype.shoot = function () {
  if (this.respawnTimer > 0) {
    return null;
  }
  return Tank.prototype.shoot.call(this, "player");
};

PlayerTank.prototype.takeDamage = function () {
  if (this.respawnTimer > 0 || this.invincibleTimer > 0) {
    return;
  }
  this.lives -= 1;
  if (this.lives > 0) {
    this.respawnTimer = CONFIG.PLAYER_RESPAWN_TIME;
    this.invincibleTimer = 0;
    this.visible = false;
  } else {
    this.kill();
  }
};

PlayerTank.prototype.respawn = function (x, y) {
  Tank.prototype.respawn.call(this, x, y);
  this.invincibleTimer = CONFIG.PLAYER_INVINCIBLE_TIME;
  this.visible = true;
  this.blinkTimer = 0;
};

PlayerTank.prototype.isDead = function () {
  return this.lives <= 0 && this.respawnTimer <= 0;
};

module.exports = PlayerTank;
